src/Juvix/Backends/Michelson/DSL/Environment.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE ConstraintKinds #-}
+
+module Juvix.Backends.Michelson.DSL.Environment where
+
+import qualified Data.Set as Set
+import qualified Juvix.Backends.Michelson.Compilation.Types as Types
+import qualified Juvix.Backends.Michelson.Compilation.VirtualStack as VStack
+import Juvix.Library hiding (show)
+import qualified Michelson.Untyped.Instr as Instr
+import qualified Michelson.Untyped.Value as V
+import Prelude (Show (..))
+
+data Env
+  = Env
+      { -- | The Virtual stack that mimics the real Michelson stack.
+        stack ∷ VStack.T Curr,
+        -- | Information during Compilation.
+        compilationLog ∷ [Types.CompilationLog],
+        -- | The Operations for the Michelson Contract.
+        ops ∷ [Types.Op],
+        -- | count of unnamed arguments for primitives.
+        count ∷ Word,
+        -- | Debug level.
+        debug ∷ Int
+      }
+  deriving (Show, Generic)
+
+type CompError = Types.CompilationError
+
+--------------------------------------------------------------------------------
+-- Top Level Types
+--------------------------------------------------------------------------------
+
+-- data Top
+--   = Curr
+--   | Completed Expanded
+
+data Expanded
+  = Constant (V.Value' Types.Op)
+  | Expanded (Instr.ExpandedOp)
+  | -- | Curr is a stand in for lambda or curry
+    Curr Curr
+  deriving (Show)
+
+newtype Fun = Fun (∀ m. Reduction m ⇒ [Types.NewTerm] → m Expanded)
+
+unFun ∷ Reduction m ⇒ Fun → [Types.NewTerm] → m Expanded
+unFun (Fun f) = f
+
+-- TODO ∷ have usage information stored with the args?
+-- May be fine without, as the terms we take ourselves should know their usage?
+data Curr
```

Comment:
As in `curried`? `current`?

Reply:
Curried

Reply:
Alright, it might be clearer to name it `data Curried` then.